lib/scholar/decomposition/incremental_pca.ex

```diff
@@ -0,0 +1,355 @@
+      """
+    ]
+  ]
+
+  tensor_opts =
+    stream_opts ++
+      [
+        batch_size: [
+          type: :pos_integer,
+          doc: "The number of samples in a batch."
+        ]
+      ]
+
+  @stream_schema NimbleOptions.new!(stream_opts)
+  @tensor_schema NimbleOptions.new!(tensor_opts)
+
+  @doc """
+  Fits an Incremental PCA model.
+
+  ## Options
+
+  #{NimbleOptions.docs(@tensor_schema)}
+  """
+  deftransform fit(%Nx.Tensor{} = x, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected input tensor to have shape {num_samples, num_features},
+             got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    opts = NimbleOptions.validate!(opts, @tensor_schema)
+
+    {num_samples, num_features} = Nx.shape(x)
+
+    batch_size =
+      if opts[:batch_size] do
+        opts[:batch_size]
+      else
+        5 * num_features
+      end
+
+    num_components = opts[:num_components]
+
+    cond do
+      num_components > batch_size ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              batch_size = #{batch_size}, got #{num_components}
+              """
+
+      num_components > num_samples ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              num_samples = #{num_samples}, got #{num_components}
+              """
+
+      num_components > num_features ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              num_features = #{num_features}, got #{num_components}
+              """
+
+      true ->
+        nil
+    end
+
+    opts = Keyword.put(opts, :batch_size, batch_size)
+
+    fit_n(x, opts)
+  end
+
+  defn fit_n(x, opts) do
+    batch_size = opts[:batch_size]
+
+    {batches, leftover} =
+      get_batches(x, batch_size: batch_size, min_batch_size: opts[:num_components])
+
+    num_batches = Nx.axis_size(batches, 0)
+    model = fit_first_n(batches[0], opts)
+
+    {model, _} =
+      while {
+              model,
+              {
+                batches,
+                i = Nx.u64(1)
+              }
+            },
+            i < num_batches do
+        batch = batches[i]
+        model = partial_fit_n(model, batch)
+        {model, {batches, i + 1}}
+      end
+
+    model =
+      case leftover do
+        nil -> model
+        _ -> partial_fit_n(model, leftover)
+      end
+
+    model
+  end
+
+  @doc """
+  Fits an Incremental PCA model on a stream of batches.
+
+  ## Options
+
+  #{NimbleOptions.docs(@stream_schema)}
+  """
+  def fit(batches = %Stream{}, opts) do
+    opts = NimbleOptions.validate!(opts, @stream_schema)
+    first_batch = Enum.at(batches, 0)
+    model = fit_first_n(first_batch, opts)
+    batches = Stream.drop(batches, 1)
+
+    Enum.reduce(
+      batches,
+      model,
+      # TODO: JIT
+      fn batch, model -> partial_fit(model, batch) end
```

Comment:
I guess this needs to be jitted?

Reply:
It will be automatically JITted once it calls `partial_fit_n` since it is a defn.

Reply:
Right, I should have thought about that. :)

Reply:
It is also completely fine to forget those things!